Let renovatebot update Go toolchain version in a single PR

tklauser opened the pull request:
Migrate the remaining Go toolchain updates from the `update-go-version` target to renovatebot and change renovatebot to group all Go version updates into a single PR.

See individual commits and #24806 for details.

Example PR: https://github.com/tklauser/cilium/pull/504

Fixes: #24806

bimmlerd reviewed (approved):
very nice. Minor nit, the first commit msg you state
> [...] GO_VERSION (currently Go 1.12.1), thus the second condition (version >= go 1.13)

where I think you meant to say "currently 1.20.1".

michi-covalent commented on .github/renovate.json5:
```diff
@@ -229,6 +232,13 @@
+        "go",
+        "docker.io/library/golang"
+      ]
+    },
```
could we do:
```suggestion
    {
      "groupName": "Go",
      "matchDepNames": [
        "go",
        "docker.io/library/golang"
      ],
      "schedule": [
        "on friday"
      ]
    },
```
image SHAs for `docker.io/library/golang` keeps changing even for the same patch version (see https://github.com/cilium/cilium/pull/24846), and renovate keeps opening new pull requests 🥺

tklauser replied:
Good point, done.

chancez reviewed (approved):
This is awesome.

aanm reviewed (approved):
:tada:

zacharysarah reviewed (commented):
LGTM from a docs perspective.

tklauser committed:
make: unconditionally set -mod=vendor for go commands

We only support building with the Go version specified in GO_VERSION
(currently Go 1.20.1), thus the second condition (version >= go 1.13) is
always true.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
ci: add renovate annotations for go-version

This allows renovate bot to update Go version in GitHub actions.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
.travis.yaml: add renovate annotations for go-version

This allows renovate bot to update the Go version used for Travis CI.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
contrib/backporting: install Go toolchain from docker image

Instead of manually downloading and installing the Go toolchain tarball,
let's just copy it from the golang image. This will allow renovate bot
to automatically update the image and thus the Go version.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
test: add renovate annotations for go-version

This allows renovate bot to update the Go version in test scripts used
within the CI VMs.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
renovate: group Go version updates

Add the necessary grouping so all Go dependencies across all files are
combined.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
tools/dev-doctor: extract minimum Go version from go.mod

Instead of having a make target update the minimum required Go version,
let dev-doctor extract it from the main module's go.mod file.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
docs: extract Go version from go.mod

The GO_VERSION file is going away in favor of renovatebot keeping the Go
toolchain version up-to-date across the tree. Let the doc generation
read the required Go major/minor version from go.mod. Note that this
will now only specify Go 1.20 instead of Go 1.20.x in the development
setup docs. This is in line with the dev-doctor tool and the
check-go-version make target also only requiring the major and minor
version to match.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
Delete GO_VERSION

All remaining users of the GO_VERSION file were either converted to be
updated by renovate or to use the go.mod file (in case only major and
minor version are required) in previous commits. Delete the now unneeded
GO_VERSION file.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
Don't enforce Go version using //go:build in main.go files

These build tags lead to a rather unhelpful error message when compiling
an outdated Go versions, e.g.

    package github.com/cilium/cilium/daemon: build constraints exclude all Go files in ...

Nothing in the error messages indicates that this is due to an outdated
Go toolchain version and this has been a source of confusion several
times for several developers.

It's also brittle in a way that if a developer adds a new tool with a
main func and forgets adding the //go:build line it will not be enforced
by tooling. The update-go-version make target will only cover existing
//go:build go1.X lines.

We have other checks in place to guarantee that Cilium is built using
the correct Go version (namely the go directive in go.mod, the
check-go-version make target, the dev-doctor tool and various CI
checks), so drop these build tags.

Signed-off-by: Tobias Klauser <tobias@cilium.io>
